cli/tools/repl.rs

```diff
@@ -211,71 +214,98 @@ impl Highlighter for Helper {
-      "#,
-      )
-      .unwrap();
-
-    Self { regex }
+    Self
   }
 }
 
+fn flatten_comments(
+  comments: SingleThreadedComments,
+) -> impl Iterator<Item = Comment> {
+  let (leading, trailing) = comments.take_all();
+  let mut comments = (*leading).clone().into_inner();
+  comments.extend((*trailing).clone().into_inner());
+  comments.into_iter().flat_map(|el| el.1)
+}
+
 impl Highlighter for LineHighlighter {
   fn highlight<'l>(&self, line: &'l str, _: usize) -> Cow<'l, str> {
-    self
-      .regex
-      .replace_all(&line.to_string(), |caps: &Captures<'_>| {
-        if let Some(cap) = caps.name("comment") {
-          colors::gray(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("string") {
-          colors::green(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("regexp") {
-          colors::red(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("number") {
-          colors::yellow(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("boolean") {
-          colors::yellow(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("null") {
-          colors::yellow(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("undefined") {
-          colors::gray(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("keyword") {
-          colors::cyan(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("infinity") {
-          colors::yellow(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("classes") {
-          colors::green_bold(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("hexnumber") {
-          colors::yellow(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("octalnumber") {
-          colors::yellow(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("binarynumber") {
-          colors::yellow(cap.as_str()).to_string()
-        } else {
-          caps[0].to_string()
-        }
-      })
-      .to_string()
-      .into()
+    let comments = SingleThreadedComments::default();
+    let lexer = Lexer::new(
+      ast::get_syntax(&MediaType::JavaScript),
+      ast::TARGET,
+      StringInput::new(line, BytePos(0), BytePos(line.len() as u32)),
+      Some(&comments),
+    );
+
+    let mut out_line = String::from(line);
+
+    // Adding color adds more bytes to the string,
+    // so an offset is needed to stop spans falling out of sync.
+    let mut offset = 0;
+
+    enum ItemKind {
+      Token(Token),
+      Comment(CommentKind),
+    }
+
+    struct Item {
+      span: Range<usize>,
+      kind: ItemKind,
+    }
+
+    fn span_to_range(span: Span) -> Range<usize> {
+      span.lo.0 as usize..span.hi.0 as usize
+    }
+
+    let mut items: Vec<Item> = Vec::new();
```

Comment:
I think this utility should be put in `cli/ast.rs` and could potenatially be reused for doc generator text output

Reply:
I've created `ast::lex()` now, assuming that's what you meant.

Reply:
Yes, that's great, thanks